### PR TITLE
Add tree-sitter based indentation

### PR DESCRIPTION
**Pull request by bbatsov**
This is a first stab at tree-sitter based indentation for erlang-ts-mode. The goal was to match erlang-mode's indentation behavior, which is also reflected in the tests -- every test case is run against both the classic erlang-mode engine and the new tree-sitter engine.

Covers all the major Erlang constructs: functions, case/if/receive/try/begin/fun blocks, records, maps, type specs, comprehensions, etc. The indentation respects `erlang-indent-level` and `erlang-indent-guard`.

Tree-sitter indentation is enabled by default, but users can toggle back to erlang-mode indentation with `M-x erlang-ts-toggle-indent-function` or by setting `erlang-ts-use-treesit-indent` to `nil`.

There's certainly room to polish things down the road, but this should be a solid foundation. Next up we can tackle structural navigation and other tree-sitter features.

### CODE REVIEW

**Comment by dgud** [comment]
Cool, I'm really picky about the indentation, though :-)

**Comment by bbatsov** [comment]
> Cool, I'm really picky about the indentation, though :-)

I expect nothing less. 😁  You can already experiment with the branch, but I marked it as "draft" because I ran into some weird CI failure on Emacs 31, and I'll need a while to troubleshoot it. It's not related to those changes, but I figured it would be good to fix it regardless.

**Comment by bbatsov** [comment]
All green! 🎉

**Comment by dgud** [comment]
Yeah I saw that as well after I merged your PR.

**Comment by garazdawi** [comment]
It would be awesome if we finally had an indenter that we can run on the whole file and get good indentation. Right now there are issues where the current indenter incorrectly indents code in code examples within triple quoted strings.

**Comment by bbatsov** [comment]
> Yeah I saw that as well after I merged your PR.

I'm reasonably sure that there's some issue in `erlang-mode` itself on Emacs 31 (which is causing the error on our CI as well, given it's a dep), but I didn't want to spend too much time on this. I hope to eventually decouple `erlang-ts-mode` completely from `erlang-mode`, so I'm not very motivated to debug `erlang-mode` at this point.

> It would be awesome if we finally had an indenter that we can run on the whole file and get good indentation. Right now there are issues where the current indenter incorrectly indents code in code examples within triple quoted strings.

Well, let's see how far I'll be able to get. 🤞 So far, so good.

**Comment by dgud** [comment]
As another test you could read the "test/resources/sample.test". 

And you can steal the ones from:
   `otp/lib/tools/test/emacs_SUITE_data/`
Where my old tests are, my idea was to add them, but if you want to continue to work on it go ahead.

**Comment by bbatsov** [comment]
I've added the OTP emacs_SUITE_data test files as you suggested - they're great for catching edge cases. Each file runs a round-trip test (strip indentation, re-indent, compare) with erlang-mode to verify correctness.

I also iterated on the tree-sitter indentation rules based on what the OTP tests revealed. The comments.erl file now passes fully with tree-sitter, and several others improved (macros, funcs, try_catch, type_specs, icr). The main improvements were: element alignment after opening delimiters, Okeefe-style function args, type spec body indentation, and the `%%%`/`%%`/`%` comment conventions.

The remaining gaps are mostly around continuation lines - cases where a multi-line expression wraps and the tree-sitter node at the beginning of the continuation line is deep inside a node that started on the previous line. These are hard to express with treesit-simple-indent-rules because the system doesn't have a native concept of "this line is a continuation." erlang-mode uses heuristics to detect this, but treesit works purely from the AST structure. I'll keep chipping away at these, but the low-hanging fruit is done.

I think it makes sense to merge this as-is and continue improving incrementally. The common cases work well, users can always toggle back to erlang-mode indentation, and the OTP test files give us a clear roadmap for what to tackle next.

**Comment by dgud** [comment]
Can you rebase, I merged the other branch. 
I rebased it but must have done something wrong because indentation didn't work at all for me.

**Comment by bbatsov** [comment]
@dgud Rebased! (I also fixed one weird interplay with the inline markdown grammar that was probably causing the indentation to misbehave for you)

**Comment by dgud** [comment]
Yes, worked better now, thanks.

@garazdawi  God morning, how should it behave no indentation at all inside doc areas?
That seems to be the only thing that I can think of.

File: erlang-ts.el [hunk 1]
```diff
@@ -470,10 +481,245 @@ Computed lazily on first use.")
...
+     ((node-is "record_field") erlang-ts--anchor-after-open-brace 0)
+     ((parent-is "record_expr") erlang-ts--anchor-after-open-brace 0)
+     ((parent-is "map_expr") erlang-ts--anchor-after-open-brace 0)
+
+     ;; Function arguments: smart alignment (Okeefe-aware)
+     ((parent-is "expr_args") erlang-ts--anchor-args 0)
+     ;; Collections: align after opening delimiter
+     ((parent-is "list") erlang-ts--anchor-after-open-delim 0)
+     ((parent-is "tuple") erlang-ts--anchor-after-open-delim 0)
+     ((parent-is "binary") erlang-ts--anchor-after-open-delim 0)
```
Comment (dgud) [inline erlang-ts.el]: "binary" regexp here matches the other binary_ .. expressions below.

**Comment by dgud** [comment]
I'm so confused the code below always says erlang, even I'm inside a doc attribute?

But it is rendered as markdown as far as I can see?

```
ELISP> (erlang-ts--language-at-point (point))
erlang
```

**Comment by garazdawi** [comment]
> @garazdawi God morning, how should it behave no indentation at all inside doc areas?
> That seems to be the only thing that I can think of.

Doc areas should have the same rules as strings, that is no indentation at all. I suppose that triple quoted strings should have some indentation rules? That is.

````erlang
foo() ->
"""
```erlang
bar() ->
ok
```
""".
````

should be indented to

````erlang
foo() ->
    """
    ```erlang
    bar() ->
    ok
    ```
    """.
````

but a single quoted string:

````erlang
foo() ->
"
```erlang
bar() ->
ok
```
".
````

should only be indented as this:

````erlang
foo() ->
    "
```erlang
bar() ->
ok
```
".
````

Another way to put it is that indentation should never ever change the meaning of the program, even if that meaning is only the documentation string.

**Comment by bbatsov** [comment]
Fixed the `"binary"` regexp issue - good catch! It's now anchored as `"^binary$"` (same for `"list"` and `"tuple"`), and I added a test for multi-line binary operations that would have caught this.

Regarding `erlang-ts--language-at-point` always returning `erlang` - that's actually the expected behavior when point is on regular Erlang code. It should only return `markdown-inline` when point is inside a `-doc` or `-moduledoc` string that the markdown-inline parser is actively covering. Could you check with `M-x treesit-explore-mode` whether the markdown-inline parser is active? You can also try `(treesit-parser-list)` in ielm to see which parsers are present - there should be both `erlang` and `markdown-inline` if the grammar is installed.

**Comment by bbatsov** [comment]
Good point about string indentation @garazdawi. I've added a rule to preserve string content during indentation - the indenter now never modifies lines inside strings, so it won't alter string values or doc attribute content (e.g., code examples inside `-doc` attributes). Added tests for both single-quoted and triple-quoted string cases.

**Comment by dgud** [comment]
It is confusing:

It is clearly rendering, markdown, I see bold text and different colors.

```
ELISP> (treesit-parser-list)
(#<treesit-parser for erlang>)
ELISP> erlang-ts--markdown-font-lock-rules
((#<treesit-compiled-query> t doc prepend))
ELISP> treesit-range-settings
((#<treesit-compiled-query> markdown-inline t nil))
```
tree-sitter explorer for array.erl
inside -doc attrib

```
forms_only: 
 (wild_attribute
  name: (attr_name - name: (atom))
  value: (string) .)
```
I'll continue tomorrow.

**Comment by bbatsov** [comment]
I took a closer look at this and to me it seems like expected behavior with `:local t` range settings. The markdown-inline parsers are created locally per range by treesit's internal machinery - they handle font-lock correctly (which is why you see bold text and colors), but they don't show up in `(treesit-parser-list)` and aren't accessible via `treesit-node-at`. So `erlang-ts--language-at-point` will always return `erlang` in this setup.

This doesn't affect any functionality - font-lock highlighting works, and indentation only uses erlang rules. The function is there for treesit's internal use when deciding which indent/font-lock rules to apply per language.

**Comment by dgud** [comment]
Ok I tried that because I was trying get the indentation engine to use the 'markdown-inline' "mode" instead of erlang,
inside `-doc` but didn't succeed.

I have played around with this, and found a couple of bugs :-) 

I don't think it is ready to set the default to use the indentation from erlang-ts yet, so if you change that I can merge
it for further work. 

Especially the `(no-node prev-line 0) ` and `ERROR` must be solved in a better way, because that is the way I have always used emacs I start with an indent to check that the code is ok so far.
I think this is doable but I have to essentially duplicate the indent rules you have made already :-(

How do you work with this, i.e. how do you reload code without starting/stopping emacs?
With things compiled/optimized/set in variables they don't get updated with a simple 'Evaluate Buffer'.

**Comment by bbatsov** [comment]
> I don't think it is ready to set the default to use the indentation from erlang-ts yet, so if you change that I can merge
it for further work.

@dgud Done!

> Especially the (no-node prev-line 0)  and ERROR must be solved in a better way, because that is the way I have always used emacs I start with an indent to check that the code is ok so far.

I get this, although I think that would be tricky as obviously when you're working with parser incomplete code fragments result in parsing errors and it's not easy to fill in the context blanks. That's not really a problem with `erlang-ts`, it's a problem with every mode using Tree-sitter. E.g. a few days ago we had a very similar conversation in clojure-ts-mode (see https://github.com/clojure-emacs/clojure-ts-mode/issues/134). 

I'll keep thinking about this, but I think it'd be nice to merge the PR in its current state, as we've made a lot of progress and it's already a pretty big PR.

**Comment by dgud** [comment]
Agree, then I also don't have to rebase all the time :-) 

Can you amend your previous commit so that the tests run with erlangs-ts-mode indentation, or is that done already?

**Comment by bbatsov** [comment]
The indentation tests are not affected by the default indentation in `erlang-ts-mode`, so no changes are needed there.

**Review by dgud — APPROVED** [review]

**Comment by dgud** [comment]
Thanks!
